### PR TITLE
add new tcp read and write methods with timeout

### PR DESCRIPTION
**Pull request by SquattingSocrates**
For db drivers to work smoothly as well as some other applications to be migrated to lunatic we should extend tcp_read and tcp_write_vectorised with timeouts. They are in separate methods for now so that we don't have to try and "encode" a None value as u64. The code in the rust-lib is also adjusted to include these new features.

### CODE REVIEW

File: crates/lunatic-networking-api/src/tcp.rs [hunk 1]
```diff
@@ -409,16 +600,80 @@ fn tcp_read<T: NetworkingCtx + ErrorCtx + Send>(
...
+    opaque_ptr: u32,
+) -> Box<dyn Future<Output = Result<u32, Trap>> + Send + '_> {
+    Box::new(async move {
+        let stream = caller
+            .data()
+            .tcp_stream_resources()
+            .get(stream_id)
+            .or_trap("lunatic::network::tcp_peek")?
+            .clone();
+        let read_timeout = stream.read_timeout.lock().await;
```
Comment (bkolobara) [inline crates/lunatic-networking-api/src/tcp.rs:650]: You are using the wrong `read_timeout` here, but it should be `peek_timeout`.

**Review by bkolobara — CHANGES_REQUESTED** [review]
Eventually we will need to find a better solution than all this mutexes. There will be no contention, but we still need to do two atomic checks and pointer chasing for each `read`, `write` or `peek` call.

I like this API interface and will merge it after the stuff below is fixed. Later we can think of a better way of handling this internally.

File: crates/lunatic-networking-api/src/tcp.rs [hunk 2]
```diff
@@ -363,23 +378,198 @@ fn tcp_write_vectored<T: NetworkingCtx + ErrorCtx + Send>(
...
+            .or_trap("lunatic::network::set_write_timeout")?
+            .clone();
+        let mut timeout = stream.write_timeout.lock().await;
+        // a way to disable the timeout
+        if duration == u64::MAX {
+            *timeout = None;
+        } else {
+            *timeout = Some(Duration::from_millis(duration));
+        }
+        Ok(0)
```
[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-networking-api/src/tcp.rs:433]: Or do you mean to return `Ok(())` ?

[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-networking-api/src/tcp.rs:433]: How would the `set_read_timeout` function look for `.wat`?
```wat
(import "lunatic::networking" "set_read_timeout" (func (param i64 i64) (result ???)))
```

[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-networking-api/src/tcp.rs]: Nevermind, already figured it out and pushed the requested changes.

**Comment by bkolobara** [comment]
Awesome! 🎉

**Pull request merged**